source/common/event/file_event_impl.h

```diff
@@ -41,5 +45,119 @@ class FileEventImpl : public FileEvent, ImplBase {
   // polling and activating new fd events.
   const bool activate_fd_events_next_event_loop_;
 };
+
+// Forward declare for friend class.
+class UserSpaceFileEventFactory;
+
+// The interface of populating event watcher and obtaining the active events. The events includes
+// Read, Write and Closed.
+class EventListener {
+public:
+  virtual ~EventListener() = default;
+
+  // Provide the activated events.
+  virtual uint32_t triggeredEvents() PURE;
+  virtual uint32_t getAndClearEphemeralEvents() PURE;
+
+  // Callbacks of the event operation.
+  virtual void onEventEnabled(uint32_t enabled_events) PURE;
+  virtual void onEventActivated(uint32_t activated_events) PURE;
+};
+
+// Return the enabled events except EV_CLOSED. This implementation is generally good since only
+// epoll supports EV_CLOSED but the entire envoy code base supports another poller. The event owner
+// must assume EV_CLOSED is never activated. Also event owner must tolerate that OS could notify
+// events which are not actually triggered.
+class DefaultEventListener : public EventListener {
+public:
+  ~DefaultEventListener() override = default;
+
+  // Return both read and write.
+  uint32_t triggeredEvents() override { return pending_events_ & (~Event::FileReadyType::Closed); }
+
+  void onEventEnabled(uint32_t enabled_events) override { pending_events_ = enabled_events; }
+
+  void onEventActivated(uint32_t activated_events) override {
+    ephermal_events_ |= activated_events;
+  }
+
+  uint32_t getAndClearEphemeralEvents() override {
+    auto res = ephermal_events_;
+    ephermal_events_ = 0;
+    return res;
+  }
+
+private:
+  // The persisted interested events and ready events.
+  uint32_t pending_events_{};
+  // The events set by activate() and will be cleared after the io callback.
+  uint32_t ephermal_events_{};
+};
+
+// A FileEvent implementation which is
+class UserSpaceFileEventImpl : public FileEvent, Logger::Loggable<Logger::Id::io> {
+public:
+  ~UserSpaceFileEventImpl() override {
+    if (schedulable_.enabled()) {
+      schedulable_.cancel();
+    }
+    ASSERT(event_counter_ == 1);
+    --event_counter_;
+  }
+
+  // Event::FileEvent
+  void activate(uint32_t events) override {
+    event_listener_.onEventActivated(events);
+    if (!schedulable_.enabled()) {
+      schedulable_.scheduleCallbackNextIteration();
+    }
+  }
+
+  void setEnabled(uint32_t events) override {
+    event_listener_.onEventEnabled(events);
+    bool was_enabled = schedulable_.enabled();
+    if (!was_enabled) {
+      schedulable_.scheduleCallbackNextIteration();
```

Comment:
Again enabled just indicates what events to watch out for.  If there's no pending data to read or write it generally doesn't trigger.